### PR TITLE
Changing matrix solver pivot element complexity

### PR DESCRIPTION
**Pull request by rath3t**
This branch already has the complexity base method but is still missing the sqrt()/tsym::sqrt() fix

### CODE REVIEW

**Comment by rath3t** [comment]
I can not see were the swaping is stored for the backtransformation at the end. because if two rows are swapped the solution at the position x(0) is then swapped to x(1) as return value. so the user gets the wrong position for the solution doesn't he?

**Comment by lubgr** [comment]
can you write a unit test showing this behavior? e.g. a simple 2x2 matrix with one diagonal element being zero. then, a pivot swap must happen, and according to your guess, the solution is incorrectly interchanged. this test should be for the old pivot finding version first, to not mix things up.

**Comment by lubgr** [comment]
Regarding your unit test: the new pivot method can be tested. You need to write a test which forces a swap of rows, i.e., a diagonal element is zero. when solving a system of equations with such a matrix, the swap must be executed. this can also be tested by locally checking the coverage (have a look at ./bin/coverage.sh). first remove all .gcda/.gcno files in ./build, then run only the single unit test and generate a coverage report. It will show you, which lines were executed during that test - the pivot-rot-swap should be covered.

**Comment by lubgr** [comment]
By the way, can you comment on what 'swapEven' means?

**Comment by rath3t** [comment]
When a swap of row or column occur the sign of the determinant of the original matrix changes. Therefore i introduces a boolean which flips everytime a swap happens and when the det is returned it checks if a even amount of swaps occured and then do nothing and if swaps are not even the returned det is -det.

**Comment by lubgr** [comment]
Isn't the determinant computed from the diagonal elements, when requested via the det() method and that's it? why is there a sign change associated with each row swap? can you give a minimal example (2x2)?

File: test/testmatrix.cpp [hunk 1]
```diff
@@ -537,11 +537,40 @@ TEST(Matrix, luDecompPivoting)
     x = A.solve(rhs);
 
     CHECK_EQUAL((4*c - 3*b)/(a*c), x(0));
-    CHECK_EQUAL(3*(b - c)/c, x(1));
+    CHECK_EQUAL(-(3*(-b + c)/c), x(1));
```
Comment (lubgr) [inline test/testmatrix.cpp:540]: 3*(b - c) is in fact the same as -(3*(-b + c)), except that the former is simpler. i guess this change is needed for the test to pass after the swapEven functionality was introduced, but this seems to be rather a hint to not do manual sign swapping, if the resulting det. is the same, only more complicated?!

File: src/matrix.cpp [hunk 2]
```diff
@@ -334,29 +332,42 @@ tsym::Vector tsym::Matrix::solveChecked(const Vector& rhs) const
...
         std::swap(data[index1][j], data[index2][j]);
+    swapEven = !swapEven;
 }
 
 void tsym::Matrix::factorizeLU()
 {
-    for (size_t j = 0; j + 1 < nCol; ++j) {
-        for (size_t i = j + 1; i < nRow; ++i) {
+    for (size_t j = 0; j + 1 < nCol; ++j){
+        for (size_t i = j + 1; i < nRow; ++i){
```
Comment (lubgr) [inline src/matrix.cpp:371]: Please be careful to not change existing formatting by accident, could you revert these changes?

File: src/matrix.cpp [hunk 3]
```diff
@@ -311,10 +311,8 @@ tsym::Vector tsym::Matrix::solveChecked(const Vector& rhs) const
     Matrix PLU(*this);
     Vector b(rhs);
     Vector x(nRow);
-
     PLU.compPartialPivots(&b);
     PLU.factorizeLU();
-
```
Comment (lubgr) [inline src/matrix.cpp]: See below. the empty lines make the method more readable, don't delete them.

File: test/testmatrix.cpp (hunk 1 above)
Comment (rath3t) [reply test/testmatrix.cpp:540]: yes otherwise the test failed. but first i look after a 2x2 example

Comment (lubgr) [reply test/testmatrix.cpp:540]: ok go for the example, but still the new expected value -(3*(-b + c))... is identical to the old one and only differs in that it's more compliated...

Comment (rath3t) [reply test/testmatrix.cpp:540]: yes somehow he didn't simplify it far enough when in the end the other sign is added?

Comment (rath3t) [reply test/testmatrix.cpp:540]: Analog unitTest

Comment (rath3t) [reply test/testmatrix.cpp:540]: Regarding the writing of a unit test of the pivot elements. The matrix after e.g. A.solve() is not change on is it? In the solvedChecked function in matrix.cpp. You do a copy of the Matrix at " Matrix PLU(*this);"? or is this just a reference? copy i think?

**Comment by rath3t** [comment]
look in the new commit were i cout everything of "simpleSymbolDet" and "tsym::Matrix::checkedDet"

**Comment by rath3t** [comment]
look in the new commit "testing 2x2matrix at simpleSymbolDet Test new" were i cout everything of "simpleSymbolDet" and "tsym::Matrix::checkedDet".
If you look at the matrix the Det should be -b*c.
After you call compPartialPivots() you get the intermediate matrix ([c d],[0,b]) then the factorization doesn't change anything and after that the intermediate matrix is passed to detFromLU() were if you only multipy the diagonal entries you end up with Det of c*b where apparently the sign is missing. Therefore the swapEven prompt was introduced. 

Now we go back to our testcomplexity.cpp and we have our normal Matrix A back were nothing was changed and therefore we can't check pivoting in a straightforward way?

**Comment by lubgr** [comment]
We really should separate concerns now, this is getting too mixed up. the PR was originaly about finding the pivot with the lowest complexit instead of taking the first non-zero one. Now we're on a completely different track that is unrelated to the way the pivot is found. Either this PR must be taken back to the original, rather simple change of selecting a pivot by complexity. Or this choice by complexity isn't included, instead the old selection mechanism should be used.

File: src/matrix.cpp [hunk 4]
```diff
@@ -440,9 +454,9 @@ tsym::Var tsym::Matrix::detFromLU() const
 
     for (size_t i = 0; i < nRow; ++i)
         det *= data[i][i];
-
+    if(swapEven==false)
+        det *= -1;
     det = det.normal();
-
```
Comment (lubgr) [inline src/matrix.cpp:445]: Please stop removing empty lines for no reason, this occurs frequently.

File: test/testmatrix.cpp [hunk 5]
```diff
@@ -653,15 +682,16 @@ TEST(Matrix, illegalLinearEqSetZeroDimension)
 
 TEST(Matrix, simpleSymbolDet)
 {
```
Comment (lubgr) [inline test/testmatrix.cpp:684]: Please don't alter an existing unit test that ensures everything keeps on working as expected (though it might fail when a bug related to the det() method is fixed, but that's also worth to be noticed about). Write a new one instead, it's only a couple of lines of code.
